<a href="https://colab.research.google.com/github/Brachaglazer/Advanced_Programming/blob/Class7_Quiz/SP_26_MCON_232_Classes_and_Objects_Quiz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# FC Classes and Objects Quiz  
**Module: Classes and Objects in Python**  

## Instructions
- This quiz is **untimed**.
- For a **fast student**, this quiz is designed to take about **60 minutes**.
- If you need to stop in order to **study**, that is allowed. However, please wait **3 hours** before resuming the quiz.
- If you need a break for any reason **other than study**, you may resume the quiz **immediately** after your break.
- **Use of AI is forbidden.**
- Please complete your own work and show what you know.

## Short reminders before you begin
A **class** is a blueprint.  
An **object** is one instance created from that blueprint.  
The `__init__` method builds the starting state of each object.  
An **instance attribute** belongs to one object, usually written with `self.attribute_name`.  
A **class attribute** belongs to the class itself and is shared unless shadowed by an instance.  
A **method** is a function inside a class.  
When an object is passed into a method, the method receives a reference to that object.  
If the method **mutates the object's attributes**, the caller sees those changes after the method ends.  
If the method **reassigns the parameter variable** to a different object, that reassignment only changes the local variable inside the method.

## Academic Integrity
By submitting this notebook, you are affirming that this work is your own and that you did not use AI or outside help that is not permitted.

## Requirements
- for any print statement `__str__` is required
- all parameters must have a data type associated with it, as well as return type from method.
- please comment ALL logic (otherwise comments are not necessary)

## Quick Reference
Keep these ideas in mind while working:

- Define a class with `class ClassName:`
- Build objects with `obj = ClassName(...)`
- Instance methods usually begin with `self`
- Access instance data with `self.attribute`
- Access class data with `ClassName.attribute`
- You can also read a class attribute through an object, but be careful: assigning through the object can create an instance attribute that shadows the class attribute
- Passing an object into another object's method allows one object to use the other object's data
- Mutation persists after the method call
- Reassignment of the parameter name does not replace the original caller's object

In [1]:

# Run this cell first.
# You may use it for quick checks throughout the quiz.

def assert_equal(actual, expected, label=""):
    if actual == expected:
        print(f"PASS: {label}")
    else:
        print(f"FAIL: {label}\n  expected: {expected}\n  got:      {actual}")

# Part A — Concepts and Vocabulary

Answer in full sentences. Keep each answer to about **2–4 sentences**.

### Question A1
What is the difference between a **class** and an **object**?

**Your answer:**
A class is the blueprint to creating specific objects. An object is an instance of a class, therefore an object is created through and belongs to a class.

### Question A2
What is the role of the `__init__` method?

**Your answer:**
The init method is the constructor for the class which initializes the attributes of the object/class.

### Question A3
What is the difference between an **instance attribute** and a **class attribute**?

**Your answer:**
An instance attribute is an attribute that holds a different value for each object created from the class, while a class attribute is an attribute that holds the same value (unless changed) for each object belonging to the class.
Meaning, the instance attribute is set when creating an object within a class, while a class attribute is defined within the class and is automatically applied to all objects created from the class. Therefore, the class attribute belongs to all objects of the class while the instance attribute is more specific to each instance/object of the class.

### Question A4
When an object is passed to another object's method, what is actually being passed?

**Your answer:**
When an object is passed to another object's method, a reference to the object is passed.

### Question A5
What is the difference between **mutating an object** and **reassigning a parameter variable** inside a method?

**Your answer:**
The difference between mutating and reassigning is that mutating changes the object itself and will therefore remain after the method ends as well. On the other hand, reassigning a parameter variable just points the variable to a new object entirely and doesn't mutate the original object at all. It can be thought of the same way as mutating vs. rebinding where rebinding creates a new object and mutating changes the original.

# Part B — Reading Code

For each question:
1. Read the code carefully.
2. Predict the output or final object state.
3. Then run the code cell if you want to check your thinking.

### Question B1 — Basic object state

Before running the code, answer:
- What are the values of `name` and `score` in `player1` after the object is created?

In [2]:

class Player:
    def __init__(self, name, score):
        self.name = name
        self.score = score

player1 = Player("Avi", 10)

print(player1.name)
print(player1.score)

Avi
10


**Your answer:**  

name: Avi  
score: 10

### Question B2 — Class variable vs instance variable

Before running the code, answer:
- What is printed?
- Why?

In [3]:

class Counter:
    total_created = 0

    def __init__(self):
        Counter.total_created += 1

a = Counter()
b = Counter()
c = Counter()

print(Counter.total_created)

3


[link text](https://)**Your answer:**  
3  
The total_created class attribute increases by one each time an object is created through this class.

### Question B3 — Passing one object into another object's method

Before running the code, answer:
- What is printed?
- Which object's data is being used?

In [4]:

class Student:
    def __init__(self, name, grade):
        self.name = name
        self.grade = grade

class Reporter:
    def show_student(self, student_obj):
        print(student_obj.name, student_obj.grade)

s1 = Student("Leah", 95)
r = Reporter()
r.show_student(s1)

Leah 95


**Your answer:**  
Leah, 95  
the s1 object of the Student class.

### Question B4 — Mutation inside a method

Before running the code, answer:
- What is printed at the end?
- Why does that change remain after the method call?

In [5]:

class BankAccount:
    def __init__(self, owner, balance):
        self.owner = owner
        self.balance = balance

class BankTeller:
    def deposit_bonus(self, account, amount):
        account.balance += amount

acct = BankAccount("Mina", 100)
teller = BankTeller()
teller.deposit_bonus(acct, 25)

print(acct.balance)

125


**Your answer:**  
125  
The change remains because a reference to the object is passed in to the method that only mutates the data and doesn't reassign/rebind it.

### Question B5 — Reassigning the parameter inside a method

Before running the code, answer:
- What is printed?
- Why does the original object stay the same?

In [6]:

class Box:
    def __init__(self, value):
        self.value = value

class Changer:
    def replace_box(self, box_obj):
        box_obj = Box(999)
        box_obj.value += 1

b = Box(10)
c = Changer()
c.replace_box(b)

print(b.value)

10


**Your answer:**  
10  
The original object stays the same because a reference to the object is passed in to the method which reassigns/rebinds so the variable points to a completely new object which is created, unconnected to the original object.

# Part C — Write the Code

Write each class or method in the code cells below.

### Question C1 — Create a simple class

Create a class named `Book`.

#### Requirements
- The constructor should accept:
  - `title`
  - `author`
  - `pages`
- Store each value in an instance attribute.
- Then create an object named `b1` with your own sample data.
- Print the title and author on one line.
- use `__str__`

**Reminder:** Each object should store its own instance data.

In [8]:

# Write your answer here
class Book:
  def __init__(self, title: str, author: str, pages: int) -> None:
    """This method is the constructor for the class which initializes the attributes"""
    self.title = title
    self.author = author
    self.pages = pages

  def __str__(self) -> str:
    """This method returns a string containing details about the object"""
    return f"{self.title} by {self.author}"

b1 = Book("Hello World", "Engineer 1", 752)
print(b1)

Hello World by Engineer 1


### Question C2 — Add an instance method

Create a class named `Dog`.

#### Requirements
- The constructor should accept:
  - `name`
  - `age`
- Store them as instance attributes.
- Add a method named `describe()` that returns a string like:
  - `"Fido is 4 years old."`
- Create one object and test your method.

**Reminder:** Instance methods use `self` to access that object's data.

In [10]:

# Write your answer here
class Dog:
  def __init__(self, name: str, age: int) -> None:
    """This method is the constructor for the class which initializes the attributes"""
    self.name = name
    self.age = age

  def describe(self) -> str:
    """This method returns a string containing details about the object"""
    return f"{self.name} is {self.age} years old"

d1 = Dog("Pup", 8)
print(d1.describe())


Pup is 8 years old


### Question C3 — Class variable practice

Create a class named `Student`.

#### Requirements
- Add a class attribute named `school_name` and set it to `"Touro"`.
- The constructor should accept:
  - `name`
  - `major`
- Store them as instance attributes.
- Create two student objects.
- Print each student's `name`.
- Print the shared class variable.

**Reminder:** The class variable belongs to the class and is shared.

In [13]:

# Write your answer here
class Student:
  school_name = "Touro"

  def __init__(self, name:str , major: str) -> None:
    """This method is the constructor for the class which initializes the attributes"""
    self.name = name
    self.major = major

st1 = Student("Bob", "Math")
st2 = Student("Joe", "Doctor")
print(st1.name)
print(st2.name)
print(Student.school_name)
#print(st1.school_name)
#print(st2.school_name)

Bob
Joe
Touro


### Question C4 — One object uses another object

Create two classes: `Player` and `Coach`.

#### `Player`
- constructor should accept:
  - `name`
  - `points`

#### `Coach`
- add a method `praise_player(player_obj)` that returns a string like:
  - `"Great job, Maya! You scored 18 points."`

Create a player object and a coach object, then call the method.

**Reminder:** This question practices passing an object into another object's method and using its data.

In [14]:

# Write your answer here
class Player:
  def __init__(self, name: str, points: int) -> None:
    """This method is the constructor for the class which initializes the attributes"""
    self.name = name
    self.points = points

class Coach:
  def praise_player(self, player_obj: Player) -> str:
    """This method returns a string using the data of the object passed in"""
    return f"Great job, {player_obj.name}! You scored {player_obj.points}."

pl1 = Player("Jon", 15)
c1 = Coach()
c1.praise_player(pl1)

'Great job, Jon! You scored 15.'

### Question C5 — Mutation vs reassignment

Create two classes: `Pet` and `Groomer`.

#### `Pet`
- constructor should accept:
  - `name`
  - `clean`

#### `Groomer`
Add two methods:

1. `wash_pet(pet_obj)`  
   - set `pet_obj.clean` to `True`

2. `replace_pet(pet_obj)`  
   - reassign `pet_obj` to a brand new `Pet("Temporary", True)`

#### After writing the classes
- Create one pet object that starts dirty.
- Call `wash_pet(...)`
- Print the original pet's `name` and `clean` value
- Then call `replace_pet(...)`
- Print the original pet's `name` and `clean` value again

#### Finally, answer in a comment:
- Which method changed the original object?
- Which method did not?
- Why?

**Reminder:** Mutation changes the object itself. Reassignment only changes the local parameter name.

In [15]:

# Write your answer here
class Pet:
  def __init__(self, name: str, clean: bool) -> None:
    """This method is the constructor for the class which initializes the attributes"""
    self.name = name
    self.clean = clean

class Groomer:
  def wash_pet(self, pet_obj: Pet) -> None:
    """This method changes the clean attribute of the pet object passed in"""
    pet_obj.clean = True

  def replace_pet(self, pet_obj: Pet) -> None:
    """This method creates a new temporary Pet object"""
    pet_obj = Pet("Temporary", True)

pet1 = Pet("Lilly", False)
g1 = Groomer()
g1.wash_pet(pet1)
print(pet1.name, pet1.clean)
g1.replace_pet(pet1)
print(pet1.name, pet1.clean)

# The wash_pet method changed the original object because it only mutated the object it had a reference to.
# The replace_pet method did not change the original object because it tried to reassign the object that it only had a reference to so it created a new object instead.

Lilly True
Lilly True


# Part D — Wrap-Up Reflection

Answer in full sentences.

### Question D1
What is one thing you understand better now about classes and objects?

### Question D2
What is still a little confusing or something you would like to practice more?

I now understand mutating and reassigning objects better.
I would like to practice class inheritance more if possible.

## Not graded :  
- How do you feel about this topic ?
- How long did this quiz take you (excluding breaks)

# Thank you !

I feel pretty comftorable with this topic and even enjoy it!  

I forgot to time myself but it took significantly quicker than the homework I did a little earlier in the day, I definitely see the progress!

Thank you!